### PR TITLE
fix(interactions): get info from cache first if available

### LINKED ISSUE #605: Status not displaying properly
### Summary

nextcord not responding with the correct Status

### Reproduction Steps

1. Enable presence/member intent
2. make a slash cmd that replies with `interaction.user.raw_status`

### Minimal Reproducible Code

```python
@client.slash_command(guild_ids=TESTING_GUILD_IDS)
async def presence(interaction: Interaction):
    await interaction.send(interaction.user.raw_status)
```

### Expected Results

show the correct status like dnd, online etc.

### Actual Results

It always displays offline whereas the user has a different status.
doing this `interaction.guild.get_member(interaction.user.id).raw_status` displays the right status but `interaction.user.raw_status` doesn't

### Intents

all

### System Information

- Python v3.9.1-final
- nextcord v2.0.0-alpha
    - nextcord pkg_resources: v2.0.0a9
- aiohttp v3.8.1
- system info: Windows 10 10.0.19041

### Checklist

- [X] I have searched the open issues for duplicates.
- [X] I have shown the entire traceback, if possible.
- [x] I have removed my token from display, if visible.

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by Soheab**
## Summary

Fixes #605

This PR changes the assignment of the following attributes for `Interaction` to try cache before constructing the object: `.message, .user`

This is needed because fields like `status` are not sent in the Interaction member payload but are usually cached.

Preview: image
I don't know how to test `.message` and User for `.user` but the correct object is returned.

## Checklist

- [x] If code changes were made then they have been tested.
  - [ ] I have updated the documentation to reflect the changes.
  - [x] I have run `task pyright` and fixed the relevant issues.
- [x] This PR fixes an issue.
- [ ] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

**Comment by DenverCoder1** [comment]
If not all intents are enabled, there may be some events that don't update the cached message I think.

Is it possible the interaction message will contain data that is newer/more accurate than the cache?

File: nextcord/interactions.py [hunk 1]
```diff
@@ -214,11 +217,15 @@ def _from_data(self, data: InteractionPayload):
             except KeyError:
                 pass
             else:
-                self.user = Member(state=self._state, guild=guild, data=member)  # type: ignore
+                cached_member = self.guild and self.guild.get_member(int(member["user"]["id"]))  # type: ignore # user key should be present here
+                self.user = cached_member or Member(state=self._state, guild=guild, data=member)  # type: ignore # user key should be present here
```
Comment (EmreTech) [inline nextcord/interactions.py:221]: I don't think it's necessary to duplicate this comment twice

**Review by EmreTech — APPROVED** [review]
Looks good

File: nextcord/interactions.py [hunk 2]
```diff
@@ -199,7 +199,10 @@ def _from_data(self, data: InteractionPayload):
 
         self.message: Optional[Message]
         try:
-            self.message = Message(state=self._state, channel=self.channel, data=data["message"])  # type: ignore
+            message = data["message"]
+            self.message = self._state._get_message(int(message["id"])) or Message(
+                state=self._state, channel=self.channel, data=message  # type: ignore
+            )
```
Comment (DenverCoder1) [inline nextcord/interactions.py:205]:
```suggestion
            message = data["message"]
            cached_message = self._state._get_message(int(message["id"]))
            if cached_message is not None:
                cached_message._update(message)
                self.message = cached_message
            else:
                self.message = Message(state=self._state, channel=self.channel, data=message)  # type: ignore
```

File: nextcord/interactions.py (hunk 1 above)
Comment (DenverCoder1) [inline nextcord/interactions.py:221]:
```suggestion
                cached_member = self.guild and self.guild.get_member(int(member["user"]["id"]))  # type: ignore # user key should be present here
                if cached_member is not None:
                    cached_member._update(member)
                    self.user = cached_member
                else:
                    self.user = Member(state=self._state, guild=guild, data=member)  # type: ignore
```

File: nextcord/interactions.py [hunk 3]
```diff
@@ -214,11 +217,15 @@ def _from_data(self, data: InteractionPayload):
...
+                cached_member = self.guild and self.guild.get_member(int(member["user"]["id"]))  # type: ignore # user key should be present here
+                self.user = cached_member or Member(state=self._state, guild=guild, data=member)  # type: ignore # user key should be present here
                 self._permissions = int(member.get("permissions", 0))
         else:
             try:
-                self.user = User(state=self._state, data=data["user"])
+                user = data["user"]
+                self.user = self._state.get_user(int(user["id"])) or User(
+                    state=self._state, data=user
+                )
```
Comment (DenverCoder1) [inline nextcord/interactions.py:228]:
```suggestion
                user = data["user"]
                cached_user = self._state.get_user(int(user["id"]))
                if cached_user is not None:
                    cached_user._update(user)
                    self.user = cached_user
                else:
                    self.user = User(state=self._state, data=user)
```

**Review by DenverCoder1 — COMMENTED** [review]
Let me know what you think of these suggestions. This should use the cache, but also update any additional fields provided by the interaction data.

**Comment by Soheab** [comment]
> If not all intents are enabled, there may be some events that don't update the cached message I think.
> 
> Is it possible the interaction message will contain data that is newer/more accurate than the cache?
> 
> Possibly they may need to be merged together?



> If not all intents are enabled, there may be some events that don't update the cached message I think.
> 
> Is it possible the interaction message will contain data that is newer/more accurate than the cache?
> 
> Possibly they may need to be merged together?


~~keep forgetting to respond, sorry~~
But no, I haven't experienced a situation where the Interaction Message was missing information or was out of date, mainly because it's not really used since Interaction has most of the stuff 

I just did it for the sake of "why not".


> If not all intents are enabled, there may be some events that don't update the cached message I think.

Same as above, I don't know but if that's true but doesn't sound good and a potential dataloss.

----

**Response to the suggestions:**
They look good to me, will of course require testing. Will do that when I can, thank you :)

**Comment by DenverCoder1** [comment]
> But no, I haven't experienced a situation where the Interaction Message was missing information or was out of date, mainly because it's not really used since Interaction has most of the stuff

I believe you are correct since if the member/message is cached at all, it should always update in the corresponding update/edit event, so just using the item from the cache should be enough.

**Review by DenverCoder1 — APPROVED** [review]

**Review by ooliver1 — APPROVED** [review]

**Pull request merged**